gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1266 @@
+        return Error.INTERNAL_ERROR.response("Error when generating JSON error response.");
+      }
+    }
+
+    private String toJson(String message) throws IOException {
+      Map<String, Object> errors = new HashMap<>();
+      errors.put("domain", domain);
+      errors.put("message", message);
+      errors.put("reason", reason);
+      Map<String, Object> args = new HashMap<>();
+      args.put("errors", ImmutableList.of(errors));
+      args.put("code", code);
+      args.put("message", message);
+      args.put("status", status);
+      return jsonFactory.toString(ImmutableMap.of("error", args));
+    }
+  }
+
+  private class RequestHandler implements HttpHandler {
+
+    private Response pickHandler(HttpExchange exchange, CallRegex regex) {
+      String path = BASE_CONTEXT.relativize(exchange.getRequestURI()).getPath();
+      String[] tokens = path.split("/");
+      String projectId = tokens.length > 0 ? tokens[0] : null;
+      String zoneName = tokens.length > 2 ? tokens[2] : null;
+      String changeId = tokens.length > 4 ? tokens[4] : null;
+      String query = BASE_CONTEXT.relativize(exchange.getRequestURI()).getQuery();
+      switch (regex) {
+        case CHANGE_GET:
+          return getChange(projectId, zoneName, changeId, query);
+        case CHANGE_LIST:
+          return listChanges(projectId, zoneName, query);
+        case ZONE_GET:
+          return getZone(projectId, zoneName, query);
+        case ZONE_DELETE:
+          return deleteZone(projectId, zoneName);
+        case ZONE_LIST:
+          return listZones(projectId, query);
+        case PROJECT_GET:
+          return getProject(projectId, query);
+        case RECORD_LIST:
+          return listDnsRecords(projectId, zoneName, query);
+        case ZONE_CREATE:
+          try {
+            return handleZoneCreate(exchange, projectId, query);
+          } catch (IOException ex) {
+            return Error.BAD_REQUEST.response(ex.getMessage());
+          }
+        case CHANGE_CREATE:
+          try {
+            return handleChangeCreate(exchange, projectId, zoneName, query);
+          } catch (IOException ex) {
+            return Error.BAD_REQUEST.response(ex.getMessage());
+          }
+        default:
+          return Error.INTERNAL_ERROR.response("Operation without a handler.");
+      }
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+      String requestMethod = exchange.getRequestMethod();
+      String rawPath = exchange.getRequestURI().getRawPath();
+      for (CallRegex regex : CallRegex.values()) {
+        if (requestMethod.equals(regex.method) && rawPath.matches(regex.pathRegex)) {
+          Response response = pickHandler(exchange, regex);
+          writeResponse(exchange, response);
+          return;
+        }
+      }
+      writeResponse(exchange, Error.NOT_FOUND.response(String.format(
+          "The url %s does not match any API call.", exchange.getRequestURI())));
```

Comment:
The URL could match but the request type could fail to match, right?  It'd be good to mention request type in the error message.